client/consensus/aura/src/collators/basic.rs

```diff
@@ -80,121 +98,136 @@ pub async fn run<Block, P, BI, CIDP, Client, RClient, SO, Proposer, CS>(
-			continue
-		}
-
-		let relay_parent_header =
-			match params.relay_client.header(RBlockId::hash(*request.relay_parent())).await {
-				Err(e) => reject_with_error!(e),
-				Ok(None) => continue, // sanity: would be inconsistent to get `None` here
-				Ok(Some(h)) => h,
+	async move {
+		let mut collation_requests = cumulus_client_collator::relay_chain_driven::init(
+			params.collator_key,
+			params.para_id,
+			params.overseer_handle,
+		)
+		.await;
+
+		let mut collator = {
+			let params = collator_util::Params {
+				create_inherent_data_providers: params.create_inherent_data_providers,
+				block_import: params.block_import,
+				relay_client: params.relay_client.clone(),
+				keystore: params.keystore.clone(),
+				para_id: params.para_id,
+				proposer: params.proposer,
+				collator_service: params.collator_service,
 			};
 
-		let claim = match collator_util::claim_slot::<_, _, P>(
-			&*params.para_client,
-			parent_hash,
-			&relay_parent_header,
-			params.slot_duration,
-			params.relay_chain_slot_duration,
-			&params.keystore,
-		)
-		.await
-		{
-			Ok(None) => continue,
-			Ok(Some(c)) => c,
-			Err(e) => reject_with_error!(e),
+			collator_util::Collator::<Block, P, _, _, _, _, _>::new(params)
 		};
 
-		let (parachain_inherent_data, other_inherent_data) = try_request!(
-			collator
-				.create_inherent_data(
-					*request.relay_parent(),
-					&validation_data,
-					parent_hash,
-					claim.timestamp(),
-				)
-				.await
-		);
-
-		let (collation, _, post_hash) = try_request!(
-			collator
-				.collate(
-					&parent_header,
-					&claim,
-					None,
-					(parachain_inherent_data, other_inherent_data),
-					// TODO [https://github.com/paritytech/cumulus/issues/2439]
-					// We should call out to a pluggable interface that provides
-					// the proposal duration.
-					Duration::from_millis(500),
-					// Set the block limit to 50% of the maximum PoV size.
-					//
-					// TODO: If we got benchmarking that includes the proof size,
-					// we should be able to use the maximum pov size.
-					(validation_data.max_pov_size / 2) as usize,
-				)
+		while let Some(request) = collation_requests.next().await {
+			macro_rules! reject_with_error {
+				($err:expr) => {{
+					request.complete(None);
+					tracing::error!(target: crate::LOG_TARGET, err = ?{ $err });
+					continue;
+				}};
+			}
+
+			macro_rules! try_request {
+				($x:expr) => {{
+					match $x {
+						Ok(x) => x,
+						Err(e) => reject_with_error!(e),
+					}
+				}};
+			}
+
+			let validation_data = request.persisted_validation_data();
+
+			let parent_header =
+				try_request!(Block::Header::decode(&mut &validation_data.parent_head.0[..]));
+
+			let parent_hash = parent_header.hash();
+
+			if !collator.collator_service().check_block_status(parent_hash, &parent_header) {
+				continue
+			}
+
+			// TODO [now]: it'd be nice to handle this gracefully when unincluded segment
+			// is at maximum.
+			match params
```

Comment:
Isn't it responsibility of `can_build_upon`?

I'm having hard time understanding "graceful" handling in this todo.

Reply:
We don't use `can_build_upon` in this collator, as it's intended to be fully backwards compatible, so doesn't depend on `AuraUnincludedSegmentApi`. Unfortunately, we can't "half-depend" on it (i.e. use it when it's present and not when it's not) because of https://github.com/paritytech/substrate/issues/13331 .

This does lead to panics when the runtime _does_ have an unincluded segment of max size 1 _and_ the collation-generation subsystem requests blocks to be built when there is a candidate pending.

Honestly, this change is unnecessary because of https://github.com/paritytech/polkadot/pull/7514 so I will just remove it.